### ISSUE TITLE
New experiments plugin - [merged]

### ISSUE THREAD

**Issue by alexander-haller** [uhd-urz/elAPI#83]
In GitLab by @mhxion on Mar 7, 2024, 03:07

_Merges add-experiments-plugin -> dev_

# Experiment plugin

## `append`

The `append` command appends text content to existing experiments. `append` accepts the following arguments.
```sh
$ elapi experiments append --help
```
![image](https://github.com/user-attachments/assets/32a9ea36-7628-4325-9463-9d85eb5db47c)

- `--id`: Experiment ID or experiment unique eLab ID.
- `--text/-t`: Text string to append to an experiment.
- `--path/-P`: Path to file (UTF-8 only) whose content to append to an experiment.
- `--markdown-to-html/-M`: If this flag is passed then `experiments` plugin attempt to convert any given `--text` or `--path` to HTML, that is if the given content is in markdown. If given content isn't in markdown then the text content is just wrapped with `<p>` tag.

We elaborate the details and a few corner cases in brief details. We use the [experiment 16](https://elabftw-dev-002.uni-heidelberg.de/experiments.php?mode=view&id=16) to demonstrate our examples.

### 1. Normal use-case

Passing any text content to `--text` is appended to the experiment without any change to the content.
> **Before:**
> ![image](https://github.com/user-attachments/assets/e11fcdc0-31e7-424a-965d-6917d0d3dc23)
```sh
$ python -m src.elapi.cli experiments append --id 16 -t "New content."
Successfully appended content to experiment.
```
> **After:**
> ![image](https://github.com/user-attachments/assets/ba6f6f02-fa40-482c-a8dc-49f7852b078b)

- Similarly, if a file is passed with `--path`, its content will be appended to the experiment. 
- If `--path` isn't UTF-8 encoded plugin will show an error and quit (exit code 1).
```sh
$ elapi experiments append --id 16 -P ~/Chemikalienverzeichnis_Beispiel.zip
ERROR    File in given path must be UTF-8 encoded!
```
- If `--path` isn't a file (but a directory) plugin will show an error and quit.
```sh
$ elapi experiments append --id 16 -P ~/Downloads
ERROR     Given path '/Users/culdesac/Downloads' must be a file!
```
- If both `--path` and `--text` are passed, the expected outcome can be ambiguous, so experiments plugin will shown an error and quit.
```sh
$ experiments append --id 16 -t "New content." -P ~/content.txt
ERROR    Either '--text/-t' or '--path/-P' can be defined, not both!
```
- If content is empty (i.e., `--text` is an empty string `""` or the `--path` file is empty), plugin will show nothing has been appended message.
```sh
$ elapi experiments append --id 16 -t ""
Content is empty. Nothing was appended to experiment.
```

### 2. Multi-line text

Passing multi-line text on the terminal via `--text` is possible depending on the terminal.

> **Before:**
> ![image](https://github.com/user-attachments/assets/d2a2eb56-aa64-4819-ad51-943699eeb331)
 ```sh
$ elapi experiments append --id 16 -t "New # <- new line
# <- new line
# <- new line
content."
```
> **After:**
> ![image](https://github.com/user-attachments/assets/ff48919c-061b-4d13-aa7b-c2749bdddbac)

### 3. Markdown to HTML

experiments plugin support converting markdown content to HTML. eLabFTW's editor supports rendering HTML text either in HTML editor mode or in markdown mode.

> **Before:**
> ![image](https://github.com/user-attachments/assets/f79fd4c3-3492-4c51-89a6-a42fdfaa4e1d)
```sh
$ python -m src.elapi.cli experiments append --id 16 -M -t "**New content.**"
Successfully appended content to experiment.
```
> **After:**
> ![image](https://github.com/user-attachments/assets/fb57d60c-680e-4257-be6b-8adec89a5ac6)

- The output of multi-line markdown content via `--text` with `-M` flag is slightly different than that of without the `-M` flag which was shown above. Multi-line here is handled better which has been already shown in [HeiChat](https://matrix-im.uni-heidelberg.de/#/room/!NLVGLfFsnkSrsDoOJT:matrix-im.uni-heidelberg.de/$McI3FYlmXZRA-fNUl7A_C9U4ce1RmEc6KitRqHTxkdo?via=matrix-im.uni-heidelberg.de).

- When eLabFTW editor is already in markdown mode, the case has been described also in [HeiChat](https://matrix-im.uni-heidelberg.de/#/room/!NLVGLfFsnkSrsDoOJT:matrix-im.uni-heidelberg.de/$MS6kmPyxJayCFe04mDiN3Upjfc3xJ2bpZs3BOXODLPQ?via=matrix-im.uni-heidelberg.de).

- `-M` flag enables the following markdown syntax features [supported](https://mistune.lepture.com/en/latest/plugins.html) by Mistune: `url`, `task_lists`, `task_lists`, `abbr`, `superscript`, `subscript`, `math`.

### Downsides of markdown conversion

One (one of many perhaps!) downside with the `-M` flag is when a file is a passed with `--path` and the file's contents aren't markdown. This, of course, isn't the intended the use-case, and mainly a mistake from the user's end. Consider the following `Chemikalienverzeichnis_Beispiel.csv` file:
```csv
4;Glutamin;Merck;L202312-04;fest;angelegt;Kategorie 1;56-85-9
```
If this CSV file was appended with `-M` (`elapi experiments append -P ./Chemikalienverzeichnis_Beispiel.csv`) then above CSV row will be parsed as:
```html
<p>4;Glutamin;Merck;L202312-04;fest;angelegt;Kategorie 1;56-85-9</p>
```
The output is simple wrapped in `<p>` tag. If the above CSV row however had _un-escaped_ markdown syntax, `-M` will try to parse the text in markdown. I.e., If `Chemikalienverzeichnis_Beispiel.csv` file contained:

```csv
4;**Glutamin**;Merck;### L202312-04;fest;angelegt;Kategorie 1;56-85-9
```

It will be parsed as:

```html
<p>4;<strong>Glutamin</strong>;Merck;### L202312-04;fest;angelegt;Kategorie 1;56-85-9</p>
     ^^^^^^^^^^^^^^^^^^^^^^^^^
```

In other words, `--markdown-to-html` is safest to use with only markdown (`.md`) and HTML files/content.

Also fixes #16.
